src/Lecture2.hs

```diff
@@ -70,7 +76,13 @@ return the removed element.
 >>> removeAt 10 [1 .. 5]
 (Nothing,[1,2,3,4,5])
 -}
-removeAt = error "TODO"
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt index list
+  | index < 0 = (Nothing, list)
+  | otherwise = go index list [] where
+    go _     [] acc = (Nothing, reverse acc)
+    go 0 (x:xs) acc = (Just x, reverse acc ++ xs)
+    go i (x:xs) acc = go (i-1) xs (x:acc)
```

Comment:
It is more efficient to construct lists using the cons operator (`:`) directly without introducing intermediate values or adding lists with `(++)` whenever possible. Adding elements to the end of the list is rather slow 🐌

`reverse` is also a rather slow function. So it's better to get the result without using it as well.

Reply:
To be honest I have no clue how to do it without intermediate list. I have simplified it to 
```
removeAt :: (Ord t, Num t) => t -> [a] -> (Maybe a, [a])
removeAt index list
  | index < 0 = (Nothing, list)
  | otherwise = go index list []
    where
      go _     []   _ = (Nothing, list)
      go 0 (x:xs) acc = (Just x, reverse acc ++ xs)
      go i (x:xs) acc = go (i-1) xs (x:acc)
```
but still have no idea how to implement expression without intermediate values.
I guess it may be done somehow like https://hackage.haskell.org/package/base-4.16.0.0/docs/src/GHC.List.html#splitAt
but reading this code it is difficult to imagine what is going on in each stack frame

Reply:
You can use recursion to extract the element and remove it from the list. So your implementation is almost correct. It just needs a little patch 🙂 

```haskell
removeAt :: (Ord t, Num t) => t -> [a] -> (Maybe a, [a])
removeAt index list
  | index < 0 = (Nothing, list)
  | otherwise = go index list
    where
      go _     [] = (Nothing, [])
      go 0 (x:xs) = (Just x, xs)
      go i (x:xs) =
          let (val, rest) = go (i - 1) xs
          in (val, x : rest)
```